### ISSUE TITLE
Use Enum for node states

### ISSUE THREAD

**Issue by gctucker** [kernelci/kernelci-pipeline#191]
The states that a Node object can go through are simply defined as strings.  Using an Enum would provide some syntaxic definition of the states and make it easier to share the same definition between the API and the client-side services.  There is already this in `api.models` which may be reused on the client sides if we can import it somehow:

```python3
class StateValues(enum.Enum):
```

At the moment, the API package is not installed in client-side services so this can't be just imported as-is.  The advantage of having it separate is that it guarantees a clear distinction between the exposed web API and its internal implementation.  One way to solve this could be to have another shared package, but that's not really practical either.  We could just recreate an enum on the client side, but it's not clear whether that's the best approach either as they would need to be manually kept in sync which is likely to fail.

**Comment by JenySadadia** [comment]
How about creating an endpoint to retrieve available node states from API?
That way the pipeline will be in sync with the API.

**Comment by gctucker** [comment]
I think that would work if it was combined with an API version, in production.  While we're still in development phase we can still implement this and have an empty API version or 0.0 until we start enforcing the version.  My point is that services can't just read the list of state names and adjust their implementation.  For example if a new state was introduced in a non-backwards compatible way the version check would ensure older clients would not use it and report an error.  It's not like user data such as the kernel revisions etc. which can be anything.

Then yes the API could return some JSON data structure which could be used to construct an Enum I suppose.  @mgalka Do you see any implementation details with a dynamic enum definition?  Maybe the state names would be fixed in the class but the strings they map to would be received from the API in an ordered list to pass as arguments.  We would then have different subclasses for when new API versions have a different list of states.

**Comment by JenySadadia** [comment]
Alternatively, to make this more generic we can return a dictionary with state names and values from API.
In the pipeline, we can construct dynamic enum with [function-call syntax](https://docs.python.org/3/library/enum.html#functional-api).
Fixed state names in the pipeline class can cause issues in case the existing state name changes in the API.

**Comment by gctucker** [comment]
The states _have_ to be fixed in the pipeline, at least for each API version, as the implementation depends on it.  How do you think services can deal with state names that change?

**Comment by JenySadadia** [comment]
Of course, services won't be able to deal with it but I suggested using state names and values from API considering that we have changed node states a few times and also added new states.
However, those changes were due to the API implementation evolving. So, yes we should have fixed state names per API version.

**Comment by JenySadadia** [comment]
I was looking into this issue and there is no way to construct a dynamic `Enum` definition.
For the sake of testing, I created an endpoint `/pending_node_states` to receive the node states in the API.
Used it in the pipeline and constructed an enum from it and stored it in a context variable.
Rather than constructing different enums as per the different API versions, we should consider defining API endpoints bound to a specific version.
We can receive a dictionary with available pending states using `/pending_node_states` and directly use it in the pipeline without the need of creating any `Enum` from it.
Looking into https://pypi.org/project/fastapi-versioning/ for it.

**Comment by gctucker** [comment]
The concept of "pending states" is only internal to the timeout services, so I don't think we should introduce things like this on the API side.  It's about how we interpret the data on the client side.

**Comment by gctucker** [comment]
About versioning, I think we'll need this in a couple of months but for now we need to make fast changes that might break services but not slow down development.  We can still put the code in place to manage versions and stick to v0.0 until we start declaring actual API versions.  That's worth a separate GitHub issue actually.

**Comment by gctucker** [comment]
Based on this discussion, I think we should just define Enums in the pipeline code with hard-coded strings based on the API documentation.  This could be done in kernelci-core (`kernelci.db.kernelci_api` as it currently stands) to have it shared across all the client-side services.  But in the end, defining strings for state names is the same as strings for endpoints: it's defined both on the API and client side, and as long as we define it in only one place on either side (so twice in practice) with ultimately an API version tied to it and the documentation is generated from the API code then things should work well.

**Comment by mgalka** [comment]
The logic of kernelci-api, hence logic of kernelci-pipeline is based on states and their transitions. So I guess that introducing versioning on the API is a must to make sure that both the serever (API) and the client (pipeline) are talking the same values. It's not only about the Enums, but also endpoint arguments and returned values. The moment we implement I'd say is "the sooner the better", but I understand there might be other priorities, so we may get a technical debt here to be paid off later.
When the versioning is in place I think we may want to create an endpoint which will serve "states enum" serialized to JSON which later can be instanciated and used over the whole client implementations.
I don't think we need minor versions for the API. In general the new version of the API should be introduced only when we break backward compatibility, so I'd rather go with `v0` rather than `v0.0`.

**Comment by JenySadadia** [comment]
Yes, the concept of "pending states" is only internal to the pipeline services, so the API endpoint should be for getting all the available node states instead of just pending states.
The pipeline services can use `kernelci.db.kernelci_api` to get all the node states and then it can define `Enum` from it as per the requirements.
Later, when we enable support for API versioning, the endpoint will be tied to the specific version.
Does this sound OK to start implementation?

**Comment by gctucker** [comment]
> When the versioning is in place I think we may want to create an endpoint which will serve "states enum" serialized to JSON which later can be instanciated and used over the whole client implementations.

Why have this for states when all the endpoints and attributes and other schema details will be defined on the client side anyway as per the documentation?

**Comment by gctucker** [comment]
I think we can already put everything in place for the version, and yes minor numbers don't matter that was just a random idea.  Then we can leave it at `v0` until we have to actually enforce it.  The reason for not publishing the API version just as yet is because it's not mature enough and doing this will add some overhead to making changes in the endpoints, schema etc. which will slow things down.

**Comment by gctucker** [comment]
Who wants to create an issue for this? (API versioning)

**Comment by mgalka** [comment]
> > When the versioning is in place I think we may want to create an endpoint which will serve "states enum" serialized to JSON which later can be instanciated and used over the whole client implementations.
> 
> Why have this for states when all the endpoints and attributes and other schema details will be defined on the client side anyway as per the documentation?

Good point. Maybe then we should think of creating the kernelci SDK that can facilitate writing kernelci client applications. But this seems like quite an effort, but still we may put it in a plan.

**Comment by mgalka** [comment]
> Who wants to create an issue for this? (API versioning)

I can take care of it. I have some issues to report today, so I won't need to switch context too much.

**Comment by gctucker** [comment]
> > > When the versioning is in place I think we may want to create an endpoint which will serve "states enum" serialized to JSON which later can be instanciated and used over the whole client implementations.
> > 
> > 
> > Why have this for states when all the endpoints and attributes and other schema details will be defined on the client side anyway as per the documentation?
> 
> Good point. Maybe then we should think of creating the kernelci SDK that can facilitate writing kernelci client applications. But this seems like quite an effort, but still we may put it in a plan.

Right, and until then we can bootstrap things by having hard-coded strings in a module in kernelci-core that all the client-side services can rely upon.  Then once we have that, the rule would be to not use any direct strings in any client implementation but only this module.

**Comment by JenySadadia** [comment]
> Who wants to create an issue for this? (API versioning)

See https://github.com/kernelci/kernelci-pipeline/issues/197
